File: sdk/python/src/ldp_protocol/types/messages.py
"""LDP message types."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Literal
from uuid import uuid4

from pydantic import BaseModel, Field

from ldp_protocol.types.error import LdpError
from ldp_protocol.types.payload import PayloadMode
from ldp_protocol.types.provenance import Provenance


class LdpMessageBody(BaseModel):
    """LDP message body — tagged union matching the Rust implementation."""

    type: str

    # HELLO
    delegate_id: str | None = None
    supported_modes: list[PayloadMode] | None = None

    # CAPABILITY_MANIFEST
    capabilities: Any | None = None

    # SESSION_PROPOSE
    config: dict[str, Any] | None = None

    # SESSION_ACCEPT
    session_id: str | None = None
    negotiated_mode: PayloadMode | None = None

    # SESSION_REJECT / SESSION_CLOSE
    reason: str | None = None

    # TASK_SUBMIT
    task_id: str | None = None
    skill: str | None = None
    input: Any | None = None

    # TASK_UPDATE
    progress: float | None = None
    message: str | None = None

    # TASK_RESULT
    output: Any | None = None
    provenance: Provenance | None = None

    # TASK_FAILED
    error: str | LdpError | None = None

    # ATTESTATION
    claim: Any | None = None
    evidence: Any | None = None

    @classmethod
    def hello(cls, delegate_id: str, supported_modes: list[PayloadMode]) -> LdpMessageBody:
        return cls(type="HELLO", delegate_id=delegate_id, supported_modes=supported_modes)

    @classmethod
    def capability_manifest(cls, capabilities: Any) -> LdpMessageBody:
        return cls(type="CAPABILITY_MANIFEST", capabilities=capabilities)

    @classmethod
    def session_propose(cls, config: dict[str, Any]) -> LdpMessageBody:
        return cls(type="SESSION_PROPOSE", config=config)

    @classmethod
    def session_accept(cls, session_id: str, negotiated_mode: PayloadMode) -> LdpMessageBody:
        return cls(type="SESSION_ACCEPT", session_id=session_id, negotiated_mode=negotiated_mode)

    @classmethod
    def session_reject(cls, reason: str | LdpError) -> LdpMessageBody:
        if isinstance(reason, str):
            error = LdpError.policy("SESSION_REJECTED", reason)
            return cls(type="SESSION_REJECT", reason=reason, error=error)
        else:
            return cls(type="SESSION_REJECT", reason=reason.message, error=reason)

    @classmethod
    def task_submit(cls, task_id: str, skill: str, input: Any) -> LdpMessageBody:
        return cls(type="TASK_SUBMIT", task_id=task_id, skill=skill, input=input)

    @classmethod
    def task_update(
        cls, task_id: str, progress: float | None = None, message: str | None = None
    ) -> LdpMessageBody:
        return cls(type="TASK_UPDATE", task_id=task_id, progress=progress, message=message)

    @classmethod
    def task_result(cls, task_id: str, output: Any, provenance: Provenance) -> LdpMessageBody:
        return cls(type="TASK_RESULT", task_id=task_id, output=output, provenance=provenance)

    @classmethod
    def task_failed(cls, task_id: str, error: str | LdpError) -> LdpMessageBody:
        if isinstance(error, str):
            error = LdpError.runtime("TASK_FAILED", error)
        return cls(type="TASK_FAILED", task_id=task_id, error=error)

    @classmethod
    def task_cancel(cls, task_id: str) -> LdpMessageBody:
        return cls(type="TASK_CANCEL", task_id=task_id)

    @classmethod
    def session_close(cls, reason: str | None = None) -> LdpMessageBody:
        return cls(type="SESSION_CLOSE", reason=reason)


class LdpEnvelope(BaseModel):
    """LDP message envelope — wraps every protocol message."""

    message_id: str = Field(default_factory=lambda: str(uuid4()))
    session_id: str = ""
    from_: str = Field(default="", alias="from")
    to: str = ""
    body: LdpMessageBody
    payload_mode: PayloadMode = PayloadMode.TEXT
    timestamp: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    provenance: Provenance | None = None
    signature: str | None = None
    signature_algorithm: str | None = None

    model_config = {"populate_by_name": True}

    @classmethod
    def create(
        cls,
        session_id: str,
        from_id: str,
        to_id: str,
        body: LdpMessageBody,
        payload_mode: PayloadMode = PayloadMode.TEXT,
    ) -> LdpEnvelope:
        return cls(
            session_id=session_id,
            **{"from": from_id},
            to=to_id,
            body=body,
            payload_mode=payload_mode,
        )

File: src/types/messages.rs
//! LDP message types.
//!
//! The LDP message envelope wraps all protocol messages with routing,
//! session context, and provenance metadata.

use crate::types::error::LdpError;
use crate::types::payload::PayloadMode;
use crate::types::provenance::Provenance;
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// LDP message envelope — wraps every protocol message.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LdpEnvelope {
    /// Message ID (UUID).
    pub message_id: String,

    /// Session ID this message belongs to.
    pub session_id: String,

    /// Sender delegate ID.
    pub from: String,

    /// Recipient delegate ID.
    pub to: String,

    /// Message body.
    pub body: LdpMessageBody,

    /// Payload mode used for this message.
    pub payload_mode: PayloadMode,

    /// ISO 8601 timestamp.
    pub timestamp: String,

    /// Optional provenance (attached to results).
    pub provenance: Option<Provenance>,

    /// HMAC signature of the message (hex-encoded).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signature: Option<String>,

    /// Signature algorithm (e.g., "hmac-sha256").
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signature_algorithm: Option<String>,
}

/// LDP message body variants.
///
/// Maps to the LDP RFC message types. DCI interaction moves are carried
/// as TASK_SUBMIT payloads (no new message types needed per integration spec).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "SCREAMING_SNAKE_CASE")]
pub enum LdpMessageBody {
    /// Initial handshake.
    Hello {
        delegate_id: String,
        supported_modes: Vec<PayloadMode>,
    },

    /// Capability manifest response.
    CapabilityManifest { capabilities: Value },

    /// Propose a session with configuration.
    SessionPropose { config: Value },

    /// Accept a proposed session.
    SessionAccept {
        session_id: String,
        negotiated_mode: PayloadMode,
    },

    /// Reject a proposed session.
    SessionReject {
        reason: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        error: Option<LdpError>,
    },

    /// Submit a task within a session.
    TaskSubmit {
        task_id: String,
        skill: String,
        input: Value,
    },

    /// Task progress update.
    TaskUpdate {
        task_id: String,
        progress: Option<f32>,
        message: Option<String>,
    },

    /// Task result.
    TaskResult {
        task_id: String,
        output: Value,
        provenance: Provenance,
    },

    /// Task failure.
    TaskFailed { task_id: String, error: LdpError },

    /// Task cancellation request.
    TaskCancel { task_id: String },

    /// Attestation (trust signal).
    Attestation { claim: Value, evidence: Value },

    /// Session close.
    SessionClose { reason: Option<String> },
}

impl LdpEnvelope {
    /// Create a new envelope with auto-generated message ID and timestamp.
    pub fn new(
        session_id: impl Into<String>,
        from: impl Into<String>,
        to: impl Into<String>,
        body: LdpMessageBody,
        payload_mode: PayloadMode,
    ) -> Self {
        Self {
            message_id: uuid::Uuid::new_v4().to_string(),
            session_id: session_id.into(),
            from: from.into(),
            to: to.into(),
            body,
            payload_mode,
            timestamp: chrono::Utc::now().to_rfc3339(),
            provenance: None,
            signature: None,
            signature_algorithm: None,
        }
    }
}
